File: 0576-out-of-boundary-paths/0576-out-of-boundary-paths.py
class Solution:
    def findPaths(self, m: int, n: int, maxMove: int, startRow: int, startColumn: int) -> int:
        
        direction = [[0,1], [1,0], [0,-1], [-1,0]]
        MOD = 10**9 + 7
        memory = [[[-1 for _ in range(maxMove + 1)] for _ in range(n)] for _ in range(m)]
        
        def solve(maxMove: int, startRow: int, startColumn: int) -> int:
            if startRow < 0 or startRow >= m or startColumn < 0 or startColumn >= n:
                return 1
            if maxMove <= 0:
                return 0
            if memory[startRow][startColumn][maxMove] != -1:
                return memory[startRow][startColumn][maxMove]
            
            result = 0
            for i in range(len(direction)):
                result = (result + solve(maxMove - 1, startRow + direction[i][0], startColumn + direction[i][1])) % MOD
            
            memory[startRow][startColumn][maxMove] = result
            return result
            
        return solve(maxMove, startRow, startColumn)

File: LeetCode/0576-out-of-boundary-paths/0576-out-of-boundary-paths.cpp
class Solution {
public:
    int MOD = 1e9+7;
    vector<vector<int>> directions{{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
    int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {

        vector<vector<int>> t(50, vector<int>(50));
        t[startRow][startColumn] = 1;
        int result = 0;

        for (int k = 1; k <= maxMove; ++k) {
            vector<vector<int>> temp(50, vector<int>(50));
            for (int i = 0; i < m; ++i) {
                for (int j = 0; j < n; ++j) {
                    
                    for(auto &dir : directions) {
                        int x = i + dir[0];
                        int y = j + dir[1];
                        if (x < 0 || x >= m || y < 0 || y >= n) {
                            result = (result + t[i][j]) % MOD;
                        }
                        else {
                            temp[x][y] = (temp[x][y] + t[i][j]) % MOD;
                        }
                    }
                }
            }
            t = temp;
        }

        return result;

    }
};